### PR TITLE
phpcs: Fix unintentional exclusions

### PR DESCRIPTION
**Pull request by anomiex**
## Proposed changes:

Turns out that if you do like
```
<rule ref="Standard.Category.Sniff.Message">
  <severity>0</severity>
</rule>
```
then phpcs may implicitly set severity to 0 for all the rest of `Standard.Category.Sniff` too since it interprets the `<rule>` as wanting only that one message from the sniff.

What you really want if you want to exclude only one message from a sniff is
```
<rule ref="Standard.Category.Sniff">
  <exclude name="Standard.Category.Sniff.Message"/>
</rule>
```

So let's add a CI linting check to tell people to do the latter when they try to do the former, and clean up all the places where people have done the former in phpcs-required files.

The cleanup takes three forms:

* Fix anything auto-fixable, because there's little point in ignoring them when they can be fixed with no effort.
* Manually fix things that are easily enough fixable and that really shouldn't be ignored.
* Add additional message exclusions for messages that it seems like they really wanted to exclude.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

None. I was looking at a plugin Phan has to flag when a doc comment is using `/*` instead of `/**`, then I thought that phpcs would probably be better for that than Phan, and then came across this when looking into why phpcs hadn't already caught these.

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* CI happy?
* You could try copying just the `.github/files/lint-project-structure.sh` from this PR into trunk and seeing if it flags all the `.phpcs.dir.xml` files I'm changing here.

### CODE REVIEW

File: .github/files/lint-project-structure.sh [hunk 1]
```diff
@@ -539,6 +539,15 @@ for FILE in $(git -c core.quotepath=off ls-files 'projects/packages/**/block.jso
 	fi
 done
 
+# - In phpcs config, `<rule ref="Standard.Category.Sniff.Message"><severity>0</severity></rule>` doesn't do what you think.
+debug "Checking for bad message exclusions in phpcs configs"
+for FILE in $(git -c core.quotepath=off ls-files .phpcs.config.xml .phpcs.xml.dist .github/files/php-linting-phpcs.xml .github/files/phpcompatibility-dev-phpcs.xml '*/.phpcs.dir.xml' '*/.phpcs.dir.phpcompatibility.xml'); do
+	while IFS=$'\t' read -r LINE REF; do
+		EXIT=1
+		echo "::error file=$FILE,line=$LINE::PHPCS config attempts to set severity 0 for the sniff message \"$REF\". To exclude a single message from a sniff, use \`<rule ref=\"${REF%.*}\"><exclude name=\"$REF\"/></rule>\` instead."
+	done < <( php -r '$doc = new DOMDocument(); $doc->load( $argv[1] ); $xpath = new DOMXPath( $doc ); function has_message( $v ) { return count( explode(".", $v[0]->value) ) >= 4; } $xpath->registerNamespace("php", "http://php.net/xpath"); $xpath->registerPHPFunctions( "has_message" ); foreach ( $xpath->evaluate( "//rule[php:function(\"has_message\", @ref)][severity[normalize-space(.)=\"0\"]]" ) as $node ) { echo "{$node->getLineNo()}\t{$node->getAttribute("ref")}\n"; }' "$FILE" )
```
Comment (tbradsha) [inline .github/files/lint-project-structure.sh]: It'd be nice to have this as a heredoc instead for readability...

File: projects/packages/wp-js-data-sync/.phpcs.dir.xml [hunk 2]
```diff
@@ -27,35 +27,30 @@
...
 	</rule>
-	<rule ref="Squiz.Commenting.FunctionComment.MissingParamName">
-		<severity>0</severity>
+	<rule ref="Generic.Commenting.DocComment">
+		<exclude name="Generic.Commenting.DocComment.MissingShort"/>
 	</rule>
-	<rule ref="Squiz.Commenting.FunctionComment.MissingParamTag">
-		<severity>0</severity>
+	<rule ref="Squiz.Commenting.FunctionComment">
+		<exclude name="Squiz.Commenting.FunctionComment.Missing"/>
```
Comment (tbradsha) [inline projects/packages/wp-js-data-sync/.phpcs.dir.xml:45]: This should be nested in the above group (line 34).

File: projects/plugins/boost/.phpcs.dir.xml [hunk 3]
```diff
@@ -21,27 +21,32 @@
...
-		<severity>0</severity>
+	<rule ref="Squiz.Commenting.FunctionComment">
+		<exclude name="Squiz.Commenting.FunctionComment.Missing"/>
+		<exclude name="Squiz.Commenting.FunctionComment.MissingParamTag"/>
+		<exclude name="Squiz.Commenting.FunctionComment.ParamCommentFullStop"/>
 	</rule>
-	<rule ref="Squiz.Commenting.FunctionComment.MissingParamComment">
-		<severity>0</severity>
+	<rule ref="Squiz.Commenting.FunctionComment">
+		<exclude name="Squiz.Commenting.FunctionComment.MissingParamComment"/>
```
Comment (tbradsha) [inline projects/plugins/boost/.phpcs.dir.xml:41]: Put this in the above group.

File: projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Action_Fix.php [hunk 4]
```diff
@@ -16,8 +16,6 @@ class Image_Analysis_Action_Fix implements Data_Sync_Action {
 	 *
 	 * @param mixed            $data    JSON Data passed to the action.
 	 * @param \WP_REST_Request $_request The request object.
-	 *
-	 * @throws \Exception
```
Comment (tbradsha) [inline projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Action_Fix.php:20]: Optionally:

```
* @return array|\WP_Error
```

File: .github/files/lint-project-structure.sh (hunk 1 above)
Comment (anomiex) [reply .github/files/lint-project-structure.sh]: Meh. Ok.

File: projects/packages/wp-js-data-sync/.phpcs.dir.xml (hunk 2 above)
Comment (anomiex) [reply projects/packages/wp-js-data-sync/.phpcs.dir.xml:45]: 🤷 Phpcs doesn't seem to mind, and the original seemed to intentionally have all the "Missing" sniffs listed as a group.

File: projects/plugins/boost/.phpcs.dir.xml (hunk 3 above)
Comment (anomiex) [reply projects/plugins/boost/.phpcs.dir.xml:41]: This one I just missed combining 😀

File: projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Action_Fix.php (hunk 4 above)
Comment (anomiex) [reply projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Action_Fix.php:20]: Meh. May as well.

**Review by tbradsha — APPROVED** [review]

**Pull request merged**